designer/client/components/Visualisation/Visualisation.tsx

```diff
@@ -1,98 +1,98 @@
-import React from "react";
-import { Data } from "@xgovformbuilder/model";
+import React, {
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from "react";
 
 import Page from "../../page";
 import { Lines } from "./Lines";
 import { Minimap } from "./Minimap";
 import { Info } from "./Info";
-import { getLayout, Layout } from "./getLayout";
+import { getLayout, Pos } from "./getLayout";
 import "./visualisation.scss";
+import { DataContext } from "../../context";
+import { i18n } from "../../i18n";
 
 type Props = {
-  data: Data;
-  id: string;
-  updatedAt: string;
-  downloadedAt: string;
-  previewUrl: string;
-  persona: any;
+  updatedAt?: string;
+  downloadedAt?: string;
+  previewUrl?: string;
+  persona?: any;
 };
 
-type State = {
-  layout?: Layout["pos"];
-};
-
-export class Visualisation extends React.Component<Props, State> {
-  ref = React.createRef<HTMLDivElement>();
-
-  constructor(props: Props) {
-    super(props);
-
-    this.state = {};
-  }
-
-  scheduleLayout() {
-    setTimeout(() => {
-      const { data } = this.props;
-      const layout = getLayout(data, this.ref.current!);
-
-      this.setState({
-        layout: layout.pos,
-      });
-    }, 200);
-  }
+export function useVisualisation(ref) {
+  const { data } = useContext(DataContext);
+  const [layout, setLayout] = useState<Pos>();
+  const [pushedHistoryCount, setPushedHistoryCount] = useState(-1);
 
-  componentDidMount() {
-    this.scheduleLayout();
-  }
+  useEffect(() => {
+    const layout = getLayout(data, ref.current!);
```

Comment:
Never seen an exclamation mark after a variable before. I've learnt something new today 😀

Reply:
yup. Since you have to init useRef with null, the type is more like "HTMLElement | null", but since you _know_ that it will be HTML element, you can use this to cast to the non-nullish type. You can do a similar thing in swift and Kotlin (they call it something like force unwrapping) :)

Reply:
Ah cool, I'll use it more often going forward.